util/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/monitoring/MonitoringService.java

```diff
@@ -0,0 +1,191 @@
+
+import com.google.api.Metric;
+import com.google.api.MetricDescriptor;
+import com.google.api.MonitoredResource;
+import com.google.api.client.http.HttpRequestFactory;
+import com.google.cloud.healthcare.imaging.dicomadapter.GcpMetadataUtil;
+import com.google.cloud.monitoring.v3.MetricServiceClient;
+import com.google.monitoring.v3.CreateTimeSeriesRequest;
+import com.google.monitoring.v3.Point;
+import com.google.monitoring.v3.ProjectName;
+import com.google.monitoring.v3.TimeInterval;
+import com.google.monitoring.v3.TimeSeries;
+import com.google.monitoring.v3.TypedValue;
+import com.google.protobuf.Timestamp;
+import com.google.protobuf.util.Timestamps;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MonitoringService {
+
+  private static final int DELAY = 60;
+
+  private static final String META_PROJECT_ID = "project/project-id";
+  private static final String META_LOCATION = "instance/zone";
+  private static final String META_CLUSTER_NAME = "instance/attributes/cluster-name";
+
+  private static Logger log = LoggerFactory.getLogger(MonitoringService.class);
+  private static MonitoringService INSTANCE;
+  private static boolean ENABLED = true;
+
+  private final MetricServiceClient client;
+  private final ScheduledExecutorService service;
+  private final HashMap<IMonitoringEvent, Long> aggregateEvents;
+  private final IMonitoringEvent[] monitoredEvents;
+  private final MonitoredResource monitoredResource;
+
+  private String projectId;
+
+  private MonitoringService(String projectId, IMonitoringEvent[] monitoredEvents,
+      HttpRequestFactory requestFactory) throws IOException {
+    client = MetricServiceClient.create();
+
+    aggregateEvents = new HashMap<>();
+
+    if (projectId != null && projectId.length() > 0) {
+      this.projectId = projectId;
+    } else {
+      this.projectId = GcpMetadataUtil.get(requestFactory, META_PROJECT_ID);
+    }
+    this.monitoredEvents = monitoredEvents;
+
+    // configure Resource
+    MonitoredResource.Builder resourceBuilder = MonitoredResource.newBuilder();
+    Map<String, String> resourceLabels = new HashMap<>();
+    resourceLabels.put("project_id", this.projectId);
+
+    Map<String, String> env = System.getenv();
+    if (env.containsKey("ENV_POD_NAME")
+        && env.containsKey("ENV_POD_NAMESPACE")
+        && env.containsKey("ENV_CONTAINER_NAME")) {
+      resourceLabels.put("pod_name", env.get("ENV_POD_NAME"));
+      resourceLabels.put("namespace_name", env.get("ENV_POD_NAMESPACE"));
+      resourceLabels.put("container_name", env.get("ENV_CONTAINER_NAME"));
+      resourceLabels.put("cluster_name", GcpMetadataUtil.get(requestFactory, META_CLUSTER_NAME));
```

Comment:
Can we fetch META_CLUSTER_NAME and META_LOCATION before here, and check that they are not null before entering into this if statement? Otherwise if there's an exception thrown in GcpMetadataUtil.get, we'll set these two attributes as null, and the k8s_container resource won't be fully configured. Instead in that scenario, it would be better to default to global resource.

Reply:
Ok